### PR TITLE
Implement parser edge case

### PR DESCRIPTION
**Pull request by ChTimTsubasa**
This PR intents to fix issue #680 that peloton would crash with statements like "SELECT *" and "SELECT;"

The reason is that the Postgres parser does not think the statements mentioned above as wrong statements and the binder and optimizer would continue to execute the wrong parse tree and eventually crash.

The solution is to explicitly check that
1. The target list is not null, which can overcome "SELECT;" 
2. The from clause list is not null **when the target contains any variables 'like * or item column'**, this would overcome "SELECT *;"

A unit test case is added in parser_test to check "SELECT *" and "SELECT;".

There may still be other edge cases in 2 since right now only constant, bool expression and expression type are checked.

### CODE REVIEW

File: src/optimizer/stats/cost.cpp [hunk 1]
```diff
@@ -74,7 +74,7 @@ void Cost::CombineConjunctionStats(const std::shared_ptr<TableStats> &lhs,
       break;
     default:
       LOG_TRACE("Cost model conjunction on expression type %d not supported",
-                type);
+                (int)type);
```
Comment (apavlo) [inline src/optimizer/stats/cost.cpp]: Change this to `LOG_WARN`.
Also use `ExpressionTypeToString()` instead of printing out the type integer value.

File: src/optimizer/stats/selectivity.cpp [hunk 2]
```diff
@@ -43,7 +43,7 @@ double Selectivity::ComputeSelectivity(const std::shared_ptr<TableStats> &stats,
       return DistinctFrom(stats, condition);
     default:
       LOG_TRACE("Expression type %d not supported for computing selectivity",
-                type);
+                (int)(condition.type));
```
Comment (apavlo) [inline src/optimizer/stats/selectivity.cpp]: Same as my other comment. Switch to `LOG_WARN` and `ExpressionTypeToString()`

File: src/parser/postgresparser.cpp [hunk 3]
```diff
@@ -184,12 +184,50 @@ parser::TableRef* PostgresParser::RangeSubselectTransform(
...
+  /* Statement like 'SELECT *;' cannot detect by postgres parser and would lead to
+   * a null list of from clause*/
+  if (root == nullptr) {
+    auto target_list = select_root->targetList;
+    // The only valid situation of a null 'from list' is that all targets are constant
+    LOG_TRACE("size is : %d", target_list->length);
+    //print_pg_parse_tree(target_list);
+    if (IsTargetListWithVariable(target_list)) {
+      throw ParserException(
+          StringUtil::Format("Error parsing SQL statement"));
```
Comment (apavlo) [inline src/parser/postgresparser.cpp]: I don't think you need `StringUtil::Format()` here.

File: src/parser/postgresparser.cpp [hunk 4]
```diff
@@ -503,6 +542,13 @@ expression::AbstractExpression* PostgresParser::FuncCallTransform(
 // It checks the type of each target and call the corresponding helpers.
 std::vector<expression::AbstractExpression*>* PostgresParser::TargetTransform(
     List* root) {
+  /* Statement like 'SELECT;' cannot detect by postgres parser and would lead to
```
Comment (apavlo) [inline src/parser/postgresparser.cpp]: Switch to double slash comments instead of `/* .. */`

File: src/parser/postgresparser.cpp [hunk 5]
```diff
@@ -503,6 +542,13 @@ expression::AbstractExpression* PostgresParser::FuncCallTransform(
...
+   * null list*/
+  if (root == nullptr) {
+      throw ParserException(
+          StringUtil::Format("Error parsing SQL statement"));
```
Comment (apavlo) [inline src/parser/postgresparser.cpp]: Same comment here about `StringUtil::Format()`

File: src/parser/postgresparser.cpp [hunk 6]
```diff
@@ -1228,8 +1274,13 @@ parser::SQLStatementList* PostgresParser::ListTransform(List* root) {
     return nullptr;
   }
   LOG_TRACE("%d statements in total\n", (root->length));
-  for (auto cell = root->head; cell != nullptr; cell = cell->next) {
-    result->AddStatement(NodeTransform((Node*)cell->data.ptr_value));
+  try {
+    for (auto cell = root->head; cell != nullptr; cell = cell->next) {
+      result->AddStatement(NodeTransform((Node*)cell->data.ptr_value));
+    }
+  } catch (ParserException e) {
```
Comment (apavlo) [inline src/parser/postgresparser.cpp]: Do we really want to catch the exception here? The error message is lost.

File: test/parser/parser_test.cpp [hunk 7]
```diff
@@ -412,6 +412,17 @@ TEST_F(ParserTests, CopyTest) {
...
+// Test that the wrong queries can be detected.
+TEST_F(ParserTests, WrongQueryTest) {
+  std::vector<std::string> queries;
+  queries.push_back("SELECT;");
+  queries.push_back("SELECT *;");
+  // Parsing
+  for (auto query : queries) {
+    UNUSED_ATTRIBUTE parser::SQLStatementList* result =
+        parser::PostgresParser::ParseSQLString(query.c_str());
+    EXPECT_EQ(result->is_valid, false);
```
Comment (apavlo) [inline test/parser/parser_test.cpp]: I think that you may want to check for `ParserException` here instead.

**Comment by ChTimTsubasa** [comment]
One of the four travis jobs is failed. 
The test failed when running the script ../script/testing/psql/psql_test.sh. The line of error causing code is 
at 
`peloton: /home/travis/build/cmu-db/peloton/src/index/bwtree_index.cpp:106: bool peloton::index::BWTreeIndex<KeyType, ValueType, KeyComparator, KeyEqualityChecker, KeyHashFunc, ValueEqualityChecker, ValueHashFunc>::CondInsertEntry(const peloton::storage::Tuple*, peloton::ItemPointer*, std::function<bool(const void*)>) [with KeyType = peloton::index::CompactIntsKey<1ul>; ValueType = peloton::ItemPointer*; KeyComparator = peloton::index::CompactIntsComparator<1ul>; KeyEqualityChecker = peloton::index::CompactIntsEqualityChecker<1ul>; KeyHashFunc = peloton::index::CompactIntsHasher<1ul>; ValueEqualityChecker = peloton::ItemPointerComparator; ValueHashFunc = peloton::ItemPointerHashFunc]: Assertion `ret == true' failed.`


The full log is as following:
```
travis_time:start:0c6582f0
[0K$ bash ../script/testing/psql/psql_test.sh
Exception Type :: Catalog
Message :: Table foo is not found
peloton: /home/travis/build/cmu-db/peloton/src/index/bwtree_index.cpp:106: bool peloton::index::BWTreeIndex<KeyType, ValueType, KeyComparator, KeyEqualityChecker, KeyHashFunc, ValueEqualityChecker, ValueHashFunc>::CondInsertEntry(const peloton::storage::Tuple*, peloton::ItemPointer*, std::function<bool(const void*)>) [with KeyType = peloton::index::CompactIntsKey<1ul>; ValueType = peloton::ItemPointer*; KeyComparator = peloton::index::CompactIntsComparator<1ul>; KeyEqualityChecker = peloton::index::CompactIntsEqualityChecker<1ul>; KeyHashFunc = peloton::index::CompactIntsHasher<1ul>; ValueEqualityChecker = peloton::ItemPointerComparator; ValueHashFunc = peloton::ItemPointerHashFunc]: Assertion `ret == true' failed.
../script/testing/psql/psql_test.sh: line 59: 15717 Aborted                 (core dumped) bin/peloton -port $PELOTON_PORT > /dev/null
***FAIL***
```

**Comment by ChTimTsubasa** [comment]
@apavlo 
The line in the tests causing error was in an insert statement:
https://github.com/cmu-db/peloton/blob/master/script/testing/dml/basic2.sql#L48
Which caused peloton not passing the insertion:
https://github.com/cmu-db/peloton/blob/master/src/index/bwtree_index.cpp#L106

I tested again using the current master and it still failed on my computer. I am afaid this error is not related to this PR.

**Comment by ChTimTsubasa** [comment]
@apavlo 

Although this time the PR has passed, all I did was only rebase on top of current master. There are still bugs in BW tree, as mentioned in #730. Other PRs could still possibly fail due to the same problem.

I have tested manually with the basic1.sql.
Peloton still crashed when inserted the new tuples at 
https://github.com/cmu-db/peloton/blob/master/script/testing/dml/basic1.sql#L51
This happens both for this PR and current master.



I
